### ISSUE TITLE
Getting delayed with Hydrofabric download

### ISSUE THREAD

**Issue by benlee0423** [CIROH-UA/NGIAB_data_preprocess#66]
Machine : M2 chip Mac OS
Took more than 1 hour, and it ask again during second run, still took 20 minutes.
Commands to reproduce:
```
git clone https://github.com/CIROH-UA/NGIAB_data_preprocess
cd NGIAB_data_preprocess
python3 -m venv .venv
source .venv/bin/activate
pip install 'ngiab_data_preprocess'
python -m ngiab_data_cli -i cat-2863657 --subset --start 2010-01-01 --end 2010-01-02 --forcings --realization
```

Message after the last command:
```
Hydrofabric is missing. Would you like to download it now? [y/n] (y): y
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/ngiab_data_cli/__main__.py", line 218, in <module>
    validate_all()
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/data_sources/source_validation.py", line 222, in 
validate_all
    validate_hydrofabric()
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/data_sources/source_validation.py", line 146, in 
validate_hydrofabric
    download_and_update_hf()
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/data_sources/source_validation.py", line 124, in 
download_and_update_hf
    download_file(hydrofabric_url, file_paths.conus_hydrofabric.with_suffix(".tar.gz"))
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/data_sources/source_validation.py", line 99, in 
download_file
    future.result() for future in futures
    ^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 401, in 
__get_result
    raise self._exception
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/benjaminlee/workspace/awi/NGIAB_data_preprocess/.venv/lib/python3.12/site-packages/data_sources/source_validation.py", line 47, in 
download_chunk
    with open(chunk_path, "wb") as f_out:
         ^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 24] Too many open files: '/Users/benjaminlee/.ngiab/hydrofabric/v2.2/conus_nextgen.tar.gz.part9'
Downloading ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 GB 582.3 kB/s  Elapsed Time: 1:01:21  Remaining Time: 0:00:00
```

**Comment by JoshCu** [comment]
huh, it looks like MacOS has a default limit of 256 for open files so it doesn't like 150 threads streaming to one file each. I'll change it to just store the values in memory rather than stream to a file

**Comment by JoshCu** [comment]
I've published a [new release](https://github.com/CIROH-UA/NGIAB_data_preprocess/releases/tag/v3.2.0) which should fix the issue. I've tested it on the mac mini and it takes about a minute to download at ~30mbs. It should be able to download at gigabit speeds from s3 but I think I'd need to rework the code to use a mix of  processes and threads to do that. Hopefully this is good enough for now, a minute's not *too* long.
